server/index.js

```diff
@@ -45,7 +45,84 @@ app.get('/api/products/:productId', (req, res, next) => {
       }
     })
     .catch(err => next(err));
+});
 
+app.get('/api/cart', (req, res, next) => {
+  if (!req.session.cartId) {
+    return res.json([]);
+  } else {
+    const sql = `select "c"."cartItemId",
+                        "c"."price",
+                        "p"."productId",
+                        "p"."image",
+                        "p"."name",
+                        "p"."shortDescription"
+                   from "cartItems" as "c"
+                   join "products" as "p" using ("productId")
+                  where "c"."cartId" = $1`;
+    const val = [req.session.cartId];
+    db.query(sql, val)
+      .then(result => res.send(result.rows));
+  }
+});
+
+app.post('/api/cart', (req, res, next) => {
+  const productId = parseInt(req.body.productId);
+  if (!productId) {
+    res.status(400).send('"productId" is required');
+  } else if (typeof productId !== 'number' || productId < 0) {
+    res.status(400).send('"productId" must be a positive number');
+  }
+  const sql = `select "price"
+                 from "products"
+                where "productId" = $1`;
+  const val = [productId];
+  db.query(sql, val)
+    .then(data => {
+      if (data.rows.length === 0) {
+        throw new ClientError(`cannot find productId ${productId}`, 400);
+      } else if (req.session.cartId) {
+        return {
+          cartId: req.session.cartId,
+          price: data.rows[0].price
+        };
+      } else {
+        const sql = `insert into "carts" ("cartId", "createdAt")
+        values (default, default)
+        returning "cartId"`;
+        return db.query(sql)
+          .then(result => {
+            return {
+              cartId: result.rows[0].cartId,
+              price: data.rows[0].price
+            };
+          });
+      }
+    })
+    .then(result => {
+      req.session.cartId = result.cartId;
+      const sql = `insert into "cartItems" ("cartId", "productId", "price")
+                        values ($1, $2, $3)
+                     returning "cartItemId"`;
+      const values = [result.cartId, productId, result.price];
+      return db.query(sql, values);
+
+    })
+    .then(result => {
+      const sql = `select "c"."cartItemId",
+                          "c"."price",
+                          "p"."productId",
+                          "p"."image",
+                          "p"."name",
+                          "p"."shortDescription"
+                     from "cartItems" as "c"
+                     join "products" as "p" using ("productId")
+                    where "c"."cartItemId" = $1`;
+      const value = [result.rows[0].cartItemId];
+      db.query(sql, value)

```

Comment:
Be sure to `return` your `db.query` in case the promise it returns is rejected. That way your final `.catch` will get the error.